### DISCUSSION TITLE
Newer IPC protocol proposal

### DISCUSSION THREAD

**Discussion by subpop** [RedHatInsights/yggdrasil#93] [Ideas]
# Overview

This proposal supersedes #51. While exploring the work involved in switching to a multi-protocol dispatch service, I encountered some shortcomings in the VARLINK implementations and found that the additional complexity introduced by multiple IPC protocols was not worth the benefit. I came to the conclusion that a single protocol is still the best approach, but wanted to replace gRPC with something more feature-rich. I knew that I eventually plan on adding a "public" D-Bus interface to `yggd`, so I decided to explore the possibility of implementing the worker/dispatcher protocol using D-Bus too.

D-Bus brings a number of advantages over gRPC. An object exported on the bus has 3 ways of interacting with it: methods, signals and properties. Methods are virtually identical to methods in a gRPC protocol, but signals and properties enable a richer interface than what gRPC offers.

## Signals

Signals are unidirectional broadcast messages that an object can emit. In this case, the "NotifyEvent" method added in #54 can be implemented natively as a signal defined in the interface specification. Clients can then add a watch handler to the bus in order to be notified when a signal is emitted.

In the case of the Dispatcher interface, a signal called "Event" will be available and will emit events when the dispatcher receives a "disconnect" command control message.

## Properties

Properties are object member values that can be exposed via the standard `org.freedesktop.DBus.Properties` interface. This interface provides a `Get`, a `Set` and a `GetAll` method to retrieve the values of a given property. Clients can watch the `org.freedesktop.DBus.Properties.PropertiesChanged` signal for changes to be notified when a property value is changed.

The Dispatcher interface will export runtime configuration values as properties, so they can be retrieved by workers at any time. The Worker interface defines a "Features" property that `yggd` will observe for changes. If a worker emits a change to this property, `yggd` will collect features about known workers and publish an updated "connection-status" message. This allows workers to republish a new set of "features" on demand.

## Development tooling

D-Bus is a well-established IPC mechanism, and as such comes with a wealth of well-developed tooling. Utilities such as `busctl` and `gdbus` can be used to inspect exported objects at runtime and can be used to invoke methods directly on the various bus objects.

# Proposed Protocol

## `com.redhat.yggdrasil.Dispatcher1`

This interface is implemented by `yggd` and is exported onto the bus with a path of `/com/redhat/yggdrasil/Dispatcher1`. The destination name `com.redhat.yggdrasil.Dispatcher1` is requested so the service can be easily reached. This is the interface through which workers will interact with the dispatcher. To invoke this method directly (assuming the object is exported on the system bus (see [Running `yggd`]() below for details)), run:

```
gdbus call --system --dest com.redhat.yggdrasil.Dispatcher1 --object-path /com/redhat/yggdrasil/Dispatcher1 --method com.redhat.yggdrasil.Dispatcher1.Transmit "echo" "1" {} []
``` 

```xml
<interface name="com.redhat.yggdrasil.Dispatcher1">
    <!-- 
        Transmit passes data through the dispatcher to the provided address.
    -->
    <method name="Transmit">
        <arg type="s" name="addr" direction="in" />
        <arg type="s" name="id" direction="in" />
        <arg type="a{ss}" name="metadata" direction="in" />
        <arg type="ay" name="data" direction="in" />

        <arg type="u" name="response_code" direction="out" />
        <arg type="a{ss}" name="response_metadata" direction="out" />
        <arg type="ay" name="response_data" direction="out" />
    </method>

    <!-- 
        Emitted by the dispatcher when certain conditions arise, such as
        unexpected network disconnections or control commands received from
        the operating service.

        The possible values for the 'name' argument are:

        1 = RECEIVED_DISCONNECT
        Emitted when the dispatcher receives the "disconnect" command.
    -->
    <signal name="Event">
        <arg type="u" name="name" />
    </signal>
</interface>
```

## `com.redhat.yggdrasil.Worker1`

This interface is implemented by a worker and exported onto the bus. It is expected that the worker exports itself onto a path that includes its "directive" value. For example, the `echo` worker exports itself onto the bus with a path of `/com/redhat/yggdrasil/Worker1/echo`. Similarly, a worker must request a name that includes the directive (i.e. `com.redhat.yggdrasil.Worker1.echo`). This path and destination name is used by the dispatcher to communicate with the worker, so if the destination or path is incorrect, the worker will not receive data. Calling `Dispatch` can be done similarly to `com.redhat.yggdrasil.Dispatcher1.Transmit`:

```bash
gdbus call --system --dest com.redhat.yggdrasil.Worker1.echo --object-path /com/redhat/yggdrasil/Worker1/echo --method com.redhat.yggdrasil.Worker1.Dispatch "echo" "1" {} []
```

```xml
<interface name="com.redhat.yggdrasil.Worker1">
    <!--
        Dispatch passes data to a worker.
    -->
    <method name="Dispatch">
        <arg type="s" name="addr" direction="in" />
        <arg type="s" name="id" direction="in" />
        <arg type="a{ss}" name="metadata" direction="in" />
        <arg type="ay" name="data" direction="in" />
    </method>

    <!-- 
        Features are a set of key/value pairs that a worker exposes for the
        dispatcher to access.
    -->
    <property name="Features" type="a{ss}" access="read" />
</interface>
```

# Example

A working implementation of this proposal can be explored on a [feature branch](https://github.com/RedHatInsights/yggdrasil/tree/worker-dbus), including a port of the "echo" worker to D-Bus.

## Running `yggd`

When starting `yggd` from this branch, the service will attempt to connect to and export its objects on the system bus. When it starts workers, they are expected to do the same thing. This allows developers to easily explore the interactions between worker and dispatcher. However, it is possible to run `yggd` on a private bus. The simplest way to do that is with the `dbus-run-session` utility.

In order to request the well-known bus name `com.redhat.yggdrasil.Dispatcher1`, a security specification file must be installed into `/etc/dbus-1/system.d/`. A basic security file is available in the `data/dbus` directory and can be installed with the `install` utility:

```bash
install -D -m 644 data/dbus/yggd.conf /etc/dbus-1/system.d/yggd.conf
```

**Comment by adamruzicka** [comment]
In Satellite, we rely on Red Hat branded RHC and non-branded yggdrasil to be able to live on the same machine and with different configurations. I don't know the ins and outs of dbus, but would this change mean that this would be no longer possible as both deployments would try to use the same name on the bus?

**Reply by subpop** [reply]
As proposed currently, yes. Both would attempt to request the same destination name. What differing purpose do you use both `rhcd` and `yggd` for?

FWIW, `yggdrasil` already exists in Fedora, and I'm planning on including it as a standalone package in RHEL 10. The `rhc` RPM will no longer include and ship a `rhcd` binary; instead it will depend on `yggdrasil` and configure `yggd` to do what `rhcd` does today.

**Reply by adamruzicka** [reply]
> What differing purpose do you use both rhcd and yggd for?

The branded one to allow direct interaction with rh cloud, the non-branded one for our internal remote execution usage. We wouldn't need two instances if a single instance could connect to multiple brokers at once.

**Reply by subpop** [reply]
The MQTT client supports multiple brokers, but it doesn't allow for differentiation of the topics based on the broker. If we were to add support for multiple brokers, I assume we'd need to support partitioning of the traffic destined for each broker across different topics too.

**Reply by subpop** [reply]
I'll keep this in mind and see if I can devise a solution to run one binary with two isolated set ups. I think it should be possible, perhaps using a templated systemd unit file to create two seperate buses, environments and configuration files.

**Reply by subpop** [reply]
Running two mosquitto brokers on my host, I was able to run the same yggdrasil code-base inside two separate `dbus-run-session` commands, directed them at the two different MQTT brokers, and they both started up, claimed the name `com.redhat.yggdrasil.Dispatcher1` on their respective buses, started up their workers (which also both claimed their names successfully). It looks like we'll be able to use the same binary to run two instances; they just need to be in separate session buses.

**Reply by subpop** [reply]
When I ran two parallel instances of yggd, they are reading from the same worker config directory (`/etc/yggdrasil/workers`) to identify and start up workers. Would you need this directory to be configurable at run time? In other words, do your two instances of yggd need to have different workers, or the same worker(s) but configured differently?

**Reply by adamruzicka** [reply]
> Would you need this directory to be configurable at run time?

Preferably yes. In worst case we could most likely live with configuring the paths at compile time, but it would be nice if we didn't have to.

> In other words, do your two instances of yggd need to have different workers, or the same worker(s) but configured differently?

Preferably different workers

**Reply by subpop** [reply]
If we're embarking on the goal of making 1 binary parallel runnable, it couldn't be at compile time. We'd need to add `worker-config-dir` as a config file/CLI flag argument so that `yggd` can read its worker config files from a different directory at runtime.

**Reply by subpop** [reply]
I added a `--worker-config-dir` option to make the directory relocatable. https://github.com/RedHatInsights/yggdrasil/pull/94

**Reply by subpop** [reply]
I added some template systemd units to the [reference branch](https://github.com/RedHatInsights/yggdrasil/tree/worker-dbus) that show how this parallelized yggd concept would work. A socket is defined in `yggdrasil-bus@.socket`. A corresponding dbus-broker service is defined in `yggdrasil-bus@.service`. Finally, a `yggd@.service` file defines a service that can run `yggd` with a unique configuration file. Assuming the instantiation name is 'bunnies', the units can be started as follows:

```
# Define an abstract UNIX domain socket named yggd_bunnies
systemctl start yggdrasil-bus@bunnies.socket

# Start a dbus-broker listening on the yggd_bunnies socket
systemctl start yggdrasil-bus@bunnies.service

# Start yggd, connecting to the yggd_bunnies socket and reading
# configuration data from /etc/yggdrasil/yggdrasil-bunnies.toml
systemctl start yggd@bunnies.service
```

**Comment by subpop** [comment]
I added a "Features" property to the `com.redhat.yggdrasil.Worker1` interface. The dispatcher will listen for changes to this property for each worker it is aware of and will trigger a republish of the `connection-status` message each time a worker changes its `Features` property.

**Reply by subpop** [reply]
I updated the [reference branch](https://github.com/RedHatInsights/yggdrasil/tree/worker-dbus) with some code that adds a match signal on the org.freedesktop.DBus.Properties.PropertiesChanged signal, as well as how to emit it from within the worker.

**Comment by subpop** [comment]
(Reposting the findings of a question I received internally about the performance differences between gRPC and D-Bus.)

I wrote a [benchmark utility](https://github.com/RedHatInsights/yggdrasil/pull/95) that publishes messages through the MQTT broker and looks for them coming out on the outbound topic. I ran the same benchmark with the gRPC IPC implementation and the D-Bus IPC implementation. Averaging the duration it took to roundtrip 100 messages with a 100ms delay between publishes:

* gRPC took 5.76ms
* D-Bus took 4.19ms

Interestingly, if I published without any delay between publishing each message, both the gRPC and D-Bus implementations exhibit an ever-increasing duration result:

```
d9544c5e-af1d-4860-9368-6f0d44a74147: 7.168941ms
1d8115fa-d9b0-4bbf-855e-b5610d08f429: 9.642947ms
931dfe90-4dc7-4555-81fb-8ee5c2117342: 12.067062ms
db86e289-2f93-49c0-a4ab-1d90504e0322: 14.752412ms
05c97c11-3d44-466a-8869-a889b0e2385a: 19.781683ms
129e358d-cc17-4ac7-98bc-41c854894c01: 20.16835ms
86853ec2-a7ac-4916-9b18-dccfb4d9b46f: 22.09152ms
4e5bc4b1-1b29-44a6-89df-a7fff98fabfa: 24.994091ms
...
```

Out of curiosity, I ran `yggd` with `dbus-run-session` (to start it with a dbus-daemon session), and I ended up with results that were faster than both _gRPC_ and _dbus-broker_.

```
b0ad65ce-47fa-4b7e-8aec-bc7d54ae2910: 7.459534ms
37748eba-5897-4a4f-afaa-d117ac33793b: 430.977µs
bbcb5ec6-04f1-4e8f-aefa-a9119f18492f: 1.046394ms
cee19e50-63a2-4a52-9bc5-566ea5c6ac68: 960.612µs
17e53357-3c34-4a83-a5ac-2df84fb4a378: 1.087019ms
4b1981fd-13f7-48a0-acc6-fa6990df8372: 807.048µs
9e57fd82-23f5-4adf-bd85-0de50c6d298e: 2.044836ms
dc1e8529-aa0f-4594-a166-45c4d68a11b4: 1.341368ms
...
```

**Comment by subpop** [comment]
Along with the change to D-Bus, I plan to include a Go package in the `worker` directory. This package will provide worker developers with a library to assist with developing a worker in Go. My goal is to reduce the amount of boilerplate code that a worker is forced to carry locally. A *very* rough preview of the package interface would look something like:

```go
// Worker implements the D-Bus com.redhat.yggdrasil.Worker1 interface and
// exports itself onto the bus using the configured directive and closure.
type Worker struct {}

// NewWorker creates a worker implementing the com.redhat.yggdrasil.Worker1
// exports it onto the bus and stores a reference to recv function, calling
// it each time a new data message is received.
func NewWorker(directive string, recv func(data yggdrasil.Data) error) {}

// Start simply blocks the calling routine until either SIGTERM or SIGINT
// is received.
func Start() error {}
```

A worker would then use this package like so:

```go

import (
    "fmt"
    "log"
    "github.com/redhatinsights/yggdrasil"
    "github.com/redhatinsights/yggdrasil/worker"
)

func main() {
    w := worker.NewWorker("echo", func(data yggdrasil.Data) error {
        fmt.Printf("got some data: %#v", data)
        return nil
    })

    if err := w.Start(); err != nil {
        log.Fatal(err)
    }
}
```

**Reply by subpop** [reply]
I added an initial implementation of this worker package to the [reference branch](https://github.com/RedHatInsights/yggdrasil/tree/worker-dbus).

**Comment by subpop** [comment]
https://github.com/RedHatInsights/yggdrasil/pull/81 is a good example of something we can fix and improve upon by exposing tags as properties on the `com.redhat.yggdrasil.Dispatcher1` interface.

**Comment by subpop** [comment]
I updated the [reference branch](https://github.com/RedHatInsights/yggdrasil/tree/worker-dbus) again with another significant change. Workers are not expected to install a "worker config" file. `yggd` no longer monitors the worker config directory (`/etc/yggdrasil/workers`) for config files. Instead, a worker may start itself however it desires. As long as it starts after `yggd` is ready, the act of claiming a well-known name on the bus will identify the worker with the dispatcher, and the two will be able to send messages. Installing a D-Bus service activation file along with your worker is encouraged, as it allows the D-Bus broker to manage worker execution for you. An example of D-Bus service activation is described in the [echo worker README](https://github.com/RedHatInsights/yggdrasil/tree/worker-dbus/worker/echo#d-bus-service-activation).

**Reply by adamruzicka** [reply]
Now that we will have the ability to start workers on our own (as opposed to having yggdrasil start them), it feels we're really close to being able to deploy yggdrasil and plugins separately in, for example, some sort of containerized scenario. I personally know very little about dbus, do you thing what I outlined above would be feasible?

**Reply by subpop** [reply]
Yea, if you run a worker in a container, I don't see why that wouldn't work as long as its able to connect to the bus address. You might need to explicitly set `DBUS_STARTER_BUS_TYPE` to "session" and set the address of the bus in `DBUS_SESSION_BUS_ADDRESS`. But as long as your worker can connect to the bus that your instance of `yggd` is connected to, you should be able to send and receive messages.

**Comment by subpop** [comment]
This proposal has now landed on `main`. :partying_face:
